index.d.ts

```diff
@@ -0,0 +1,81 @@
+export interface Action {
+  type: string;
+}
+
+
+/* reducers */
+
+export type Reducer<S> = <A extends Action>(state: S, action: A) => S;
+
+export function combineReducers<S>(reducers: {[key: string]: Reducer<any>}): Reducer<S>;
+
+
+/* store */
+
+export interface Dispatch {
+  <A>(action: A): A;
+  <A, B>(action: A): B;
+}
+
+export interface Store<S> {
+  dispatch: Dispatch;
+  getState(): S;
+  subscribe(listener: () => void): () => void;
+  replaceReducer(reducer: Reducer<S>): void;
+}
+
+export interface StoreCreator {
+  <S>(reducer: Reducer<S>): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S): Store<S>;
+
+  <S>(reducer: Reducer<S>, enhancer: StoreEnhancer): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S, enhancer: StoreEnhancer): Store<S>;
+}
+
+export type StoreEnhancer = (next: StoreCreator) => StoreCreator;
+
+export const createStore: StoreCreator;
+
+
+/* middleware */
+
+export interface MiddlewareAPI<S> {
+  dispatch: Dispatch;
+  getState(): S;
+}
+
+export interface Middleware {
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch) => <A, B>(action: A) => B;
+}
+
+export function applyMiddleware(...middlewares: Middleware[]): StoreEnhancer;
+
+
+/* action creators */
+
+export interface ActionCreator {
+  <O>(...args: any[]): O;
+}
+
+export function bindActionCreators<
+  T extends ActionCreator|{[key: string]: ActionCreator}
+>(actionCreators: T, dispatch: Dispatch): T;
+
+
```

Comment:
## Action creators:

I think that overloading is much better than intersection:

```typescript
// or maybe "<A  extends Action>" but due to middleware issues better without "extends Action"
interface ActionCreator<A> extends Function {
    (...args: any[]): A;
}

interface ActionCreatorsMapObject {
    [key: string]: <A>(...args: any[]) => A;
}

interface ActionCreatorsBinder {
    <A>(actionCreator: ActionCreator<A>, dispatch: Dispatch): ActionCreator<A>;
    <M extends ActionCreatorsMapObject>(actionCreators: M, dispatch: Dispatch): M;
    (actionCreators: ActionCreatorsMapObject, dispatch: Dispatch): ActionCreatorsMapObject;
}
```

I've tested it with such piece of useless code:

```typescript

const oneCreatorA = bindActionCreators<Action>(a => a, store.dispatch);
const oneCreatorB = bindActionCreators(oneCreatorA, store.dispatch);
const oneCreatorC = bindActionCreators(oneCreatorB, store.dispatch);

interface Ott extends ActionCreatorsMapObject {
    one: IActionCreator<Action>;
    two: IActionCreator<Action>;
    three: IActionCreator<Action>;
    four: IActionCreator<Action>;
}

const xCreatorA = bindActionCreators({
    one: a => a,
    two(b: Action) {return b;},
    three: oneCreatorA
}, store.dispatch);

const cMap: Ott = {
    one: a => a,
    two(b: Action) {return b;},
    three: oneCreatorA,
    four: oneCreatorC
};

const xCreatorB: ActionCreatorsMapObject = bindActionCreators(xCreatorA, store.dispatch);
const xCreatorC: Ott = bindActionCreators(cMap, store.dispatch);
```